### PR TITLE
ENH: Add engine="numba" to groupby mean

### PR DESCRIPTION
**Pull request by mroeschke**
- [x] tests added / passed
- [x] Ensure all linting tests pass, see [here](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit) for how to run them
- [x] whatsnew entry

### CODE REVIEW

File: pandas/core/groupby/groupby.py [hunk 1]
```diff
@@ -1251,6 +1252,37 @@ def _numba_prep(self, func, data):
...
+        if cache_key not in NUMBA_FUNC_CACHE:
+            NUMBA_FUNC_CACHE[cache_key] = aggregator
+
+        index = self.grouper.result_index
+        if data.ndim == 1:
+            result_kwargs = {"name": data.name}
+            result = result.ravel()
+        else:
+            result_kwargs = {"columns": data.columns}
+        return data._constructor(result, index=index, **result_kwargs)
```
Comment (jbrockmendel) [inline pandas/core/groupby/groupby.py:1284]: need to be careful about 1) self.axis, 2) self.as_index, 3) self.observed.  should be able to re-use one of the existing result-wrapping methods to handle all these appropriately

Comment (mroeschke) [reply pandas/core/groupby/groupby.py:1284]: Fair point. Do you have any suggestions on which result-wrapping method I should use where this method passes a 1 or 2D numpy array?

Comment (jbrockmendel) [reply pandas/core/groupby/groupby.py:1284]: none of them take ndarrays directly.  my suggestion would be to make the numba function into a blk_func that gets passed to mgr.grouped_reduce, and then gets wrapped with _wrap_agged_manager

File: pandas/core/groupby/groupby.py [hunk 2]
```diff
@@ -1781,6 +1818,23 @@ def mean(self, numeric_only: bool | lib.NoDefault = lib.no_default):
             Include only float, int, boolean columns. If None, will attempt to use
             everything, then use only numeric data.
 
+        engine : str, default None
```
Comment (jbrockmendel) [inline pandas/core/groupby/groupby.py:1821]: im not wild about adding to the API.  is there a viable way to just detect if numba is available and if so always use it?

Comment (mroeschke) [reply pandas/core/groupby/groupby.py:1821]: `maybe_use_numba` can detect if the user has specified the numba option as a `config` option. 

Though, these keyword options are consistent with the reduction methods on windows ops and my plan to also expose this regular `DataFrame/Series` reduction methods as well.

Comment (jreback) [reply pandas/core/groupby/groupby.py:1821]: OT: should make this into a template that can append rather than copy-pasting

**Review by jreback — CHANGES_REQUESTED** [review]

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
rebase to make sure typing is ok (merged the typing numba PR).

**Comment by jreback** [comment]
thanks @mroeschke

**Pull request merged**